gh-104745: Limit starting a patcher more than once without stopping it

Red4Ru opened the pull request:
Previously, this would cause an `AttributeError` if the patch stopped more than once after this, and would also disrupt the original patched object.


<!-- gh-issue-number: gh-104745 -->
* Issue: gh-104745
<!-- /gh-issue-number -->

Red4Ru committed:
Limit starting a patcher more than once without stopping it

Previously, this would cause an `AttributeError` if the patch stopped more than once after this, and would also disrupt the original patched object.

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
-        self.temp_original = original
-        self.is_local = local
-        self._exit_stack = contextlib.ExitStack()
+        self._started_context = _PatchStartedContext(
```
I think the naming `self._context` would be sufficient.

picnixz replied:
```suggestion
		exit_stack = contextlib.ExitStack()
        self._started_context = _PatchStartedContext(
```

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
+            exit_stack=contextlib.ExitStack(),
+            is_local=is_local,
+            target=self.getter(),
+            temp_original=original,
```
This attribute can be called `original` since we are anyway wrapping it in a new object.

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1320,6 +1320,14 @@ def _check_spec_arg_typos(kwargs_to_check):
             )
 
 
+class _PatchStartedContext(object):
```
To reduce the memory footprint, how about using a namedtuple? I would also call it only `class _PatchContext` because it does not really give more information to have the "Started" IMO.

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
+        self._started_context = _PatchStartedContext(
+            exit_stack=contextlib.ExitStack(),
+            is_local=is_local,
+            target=self.getter(),
```
```suggestion
            target=target,
```

You may not necessarily want to re-call self.getter() which could cause a side-effect.

Red4Ru replied:
Oh, good one! I refactored a bit and overlooked this after some copy-pasting

Red4Ru replied:
Interesting that `self.getter()` is also called in `get_original` method. Should I also add optional argument in here..?

picnixz replied:
Let's try to keep the same behaviour as before unless there is a need to change.

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
                     extra_args[self.attribute_name] =  new
                 for patching in self.additional_patchers:
-                    arg = self._exit_stack.enter_context(patching)
+                    arg = self._started_context.exit_stack.enter_context(patching)
```
```suggestion
                    arg = exit_stack.enter_context(patching)
```

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1600,17 +1627,20 @@ def __enter__(self):
-        self.is_local = local
-        self._exit_stack = contextlib.ExitStack()
+        self._started_context = _PatchStartedContext(
+            exit_stack=contextlib.ExitStack(),
```
```suggestion
            exit_stack=exit_stack,
```

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1469,13 +1478,31 @@ def get_original(self):
         return original, local
 
+    @property
+    def is_started(self):
```
Previously, those were writable attributes and now it's no more the case. Could there be some code in the wild assuming so? (for instance pytest which makes quite hacky things, though I don't know if they do hacky things with this specific part of CPython).

Red4Ru replied:
I guess so... I have committed property setters just in case. Will write tests if needed when we figure out what to do with `temp_original`: do we preserve it and somehow deprecate or anything else

Red4Ru replied:
Fixed them. A bit ugly but anyway these setters exist not for an intended usecase but for backwards compatibility only

ZeroIntensity commented on Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst:
```diff
@@ -0,0 +1,5 @@
+:func:`unittest.mock.patch.object` or :func:`unittest.mock.dict`) more than
+once without stopping it. Previously, this would cause an
+:exc:`AttributeError` if the patch stopped more than once after this, and
+would also disrupt the original patched object.
```
NEWS entries should be pretty short:
```suggestion
Limit starting a patcher (from :func:`unittest.mock.patch`,
:func:`unittest.mock.patch.object` or :func:`unittest.mock.patch.dict`) more than
once without stopping it.
```

picnixz replied:
It's actually `unittest.mock.patch.dict`: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch.dict

ZeroIntensity replied:
Oh! Thank you!

picnixz commented on Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst:
```diff
@@ -0,0 +1,5 @@
+Limit starting a patcher (from :func:`unittest.mock.patch`,
+:func:`unittest.mock.patch.object` or :func:`unittest.mock.dict`) more than
```
```suggestion
:func:`unittest.mock.patch.object` or :func:`unittest.mock.patch.dict`) more than
```

Red4Ru committed:
Update Misc/NEWS.d/next/Library/2024-11-10-18-14-51.gh-issue-104745.zAa5Ke.rst

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>

Red4Ru committed:
Refactor patch context

Red4Ru committed:
Add property setters (untested)

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1469,13 +1474,51 @@ def get_original(self):
+
+    @original.setter
+    def original(self, value):
+        self._context.original = value
```
Actually when I said to rename it to `original`, it was only for the `context`'s field, not for the patch object itself!

For the patch's object, I think we can keep `temp_original` (or maybe you wanted to have the same name for both the context dataclass and the patch object? if so, sorry for the back-and-forth)

Red4Ru replied:
Agreed, let's keep `temp_original` in `_patch` while renaming it for `._context`

Red4Ru committed:
Avoid re-call of self.getter

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

Red4Ru committed:
Preserve only _patch.temp_original instead of _patch.original

Red4Ru committed:
Change NEWS file

I rechecked and get that `unittest.mock.patch.dict` is not involved. Also deleted a trailing comma

Red4Ru committed:
Fix of property setters along with a test for them

picnixz commented on Lib/unittest/mock.py:
```diff
@@ -1469,13 +1474,58 @@ def get_original(self):
+            is_local=value,
+            original=self._context.original,
+            target=self._context.target,
+        )
```
Urgh, I forgot that you cannot change the value of namedtuples. Ok, my suggestion using namedtuples was wrong. To reduce memory footprint, we can use `__slots__` in a regular class instead like you had before. That way, we save an `from collections import namedtuple` as well and simplify the property's setter. WDYT? (again sorry for this bad suggestion).

cjw296 reviewed (commented):
Can you explain why we need the whole `_PatchContext` thing? 
What couldn't be achieved by just having a `self.is_started` boolean?

cjw296 reviewed (changes requested):
This feels like we could just add an `is_started` attribute, which is set to `True` when the patcher is started and `False` when it is stopped.

I'd prefer to see that smaller, simpler change, tbh. 

But, great to see the new tests in this area!

Red4Ru committed:
Refactor _PatchContext from namedtuple to plain class with __slots__

Red4Ru committed:
Minimize changes in mock.py

cjw296 commented on Lib/test/test_unittest/testmock/testpatch.py:
```diff
@@ -1098,7 +1146,7 @@ def test_new_callable_patch(self):
         self.assertIsNot(m1, m2)
         for mock in m1, m2:
-            self.assertNotCallable(m1)
+            self.assertNotCallable(mock)
```
Nice spot!

cjw296 commented on Lib/unittest/mock.py:
```diff
@@ -1491,7 +1494,7 @@ def __enter__(self):
             raise TypeError("Can't provide explicit spec_set *and* spec or autospec")
 
-        original, local = self.get_original()
+        original, is_local = self.get_original()
```
Please undo this change.

cjw296 commented on Lib/unittest/mock.py:
```diff
@@ -1601,8 +1604,9 @@ def __enter__(self):
 
         self.temp_original = original
-        self.is_local = local
+        self.is_local = is_local
```
Please undo this change.

cjw296 commented on Lib/unittest/mock.py:
```diff
@@ -1633,11 +1640,12 @@ def __exit__(self, *exc_info):
         del self._exit_stack
+        del self.is_local
+        del self.temp_original
+        del self.target
```
Please undo this change: `self.is_started = False` should be just before the return, and that feels like it should be the only change here.

Red4Ru replied:
Undone all 3

Red4Ru committed:
Leave as little changes in mock.py as possible
